### PR TITLE
[CI] - NPM Dependency Caching and Lighthouse Audit Action

### LINKED ISSUE #2773: [CI] Improve build and deploy time
**Current Behavior**
Currently, it takes about 8 minutes for this site's [build and preview workflow](https://github.com/layer5io/layer5/actions/workflows/build-and-preview-site.yml) to execute.

**Desired Situation**
Ideally, this time is cut in half.

---
**Contributor Resources**

The layer5.io website uses Gatsby, React, and GitHub Pages. Site content is found under the [`master` branch](https://github.com/layer5io/layer5/tree/master).
- See [contributing instructions](https://github.com/layer5io/layer5/blob/master/CONTRIBUTING.md)
- See Layer5 site designs in this [Figma project](https://www.figma.com/file/5ZwEkSJwUPitURD59YHMEN/Layer5-Designs). Join the [Layer5 Community](http://slack.layer5.io) for access.

### PR DESCRIPTION
**Pull request by adithyaakrishna**
**Description**

This PR fixes #2773 

**Notes for Reviewers**

- Added NPM Caching Capabilities - This is working as of now as I had worked on a similar issue for my personal portfolio repositiory. Here's the deployment script, https://github.com/adithyaakrishna/adithyaakrishna.github.io/blob/source/.github/workflows/deploy.yml  and in this the build time has been reduced from 2-3mins to 1.5min. This successfully caches and uses it in the action when it's triggered.

**Ref:** https://github.com/adithyaakrishna/adithyaakrishna.github.io/actions - Only the two most recent action triggers
Screenshot 2022-07-05 at 1 10 32 AM

- Added Lighthouse Audit GitHub Action - WIP

**[Signed commits](https://github.com/layer5io/layer5/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: .github/workflows/lighthouse.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        uses: jakepartusch/wait-for-netlify-action@v1
+        id: netlify
+        with:
+          site_name: 'layer5'
+
+      - name: Audit URLs using Lighthouse for desktop
+        # https://github.com/treosh/lighthouse-ci-action
+        uses: treosh/lighthouse-ci-action@v9
+        with:
+          urls: ${{ steps.netlify.outputs.url }}
```
Comment (adithyaakrishna) [inline .github/workflows/lighthouse.yml:26]: @Nikhil-Ladha is there any way using which we can share outputs from other workflows? I am a bit stuck on that here 😶 

CC @Chadha93

**Comment by l5io** [comment]
🚀 Preview for commit 27a1d1a0f1d056c873d677603d18f5b9d806b50c at: https://62c34e83617947399fca538d--layer5.netlify.app

**Comment by leecalcote** [comment]
Oh! This is exciting. 😃

File: .github/workflows/build-and-deploy-site.yml [hunk 2]
```diff
@@ -8,12 +8,28 @@ jobs:
...
+      - name: Check for NPM Cache ✅
+        uses: actions/cache@v3
+        id: npm-cache # use this to check for `cache-hit` ==> if: steps.npm-cache.outputs.cache-hit != 'true'
+        with:
+          path: ${{ steps.npm-cache-dir.outputs.dir }}
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
+
+      - name: Install NPM Dependencies 👀 # Checks yarn cache and use it for dependencies install
```
Comment (Nikhil-Ladha) [inline .github/workflows/build-and-deploy-site.yml:27]:
```suggestion
      - name: Install NPM Dependencies 👀 # Checks npm cache and use it for dependencies install
```

File: .github/workflows/build-and-preview-site.yml [hunk 3]
```diff
@@ -10,23 +10,40 @@ jobs:
...
+        uses: actions/cache@v3
+        id: npm-cache # use this to check for `cache-hit` ==> if: steps.npm-cache.outputs.cache-hit != 'true'
+        with:
+          path: ${{ steps.npm-cache-dir.outputs.dir }}
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
 
-      - name: Zip Site
+      - name: Install NPM Dependencies 👀 # Checks yarn cache and use it for dependencies install
```
Comment (Nikhil-Ladha) [inline .github/workflows/build-and-preview-site.yml:31]: Same as above

**Comment by Nikhil-Ladha** [comment]
I was thinking about this PR, and wondered is it actually beneficial to have?
For us dependency installation hardly takes a minute, it is the site build that takes most of the time. And even after caching the npm dependencies that are not going to reduce.

We need an upgrade to v4, which has been pending a long time on me and if someone can take that up, that should help us make things faster (if not much). Because, the limitation here is GitHub actions env, which is very basic and no matter the big improvement locally, when it comes to GitHub actions the performance improvement becomes negligible.

**Comment by adithyaakrishna** [comment]
@Nikhil-Ladha I can take this up as I had worked on the v2 to v3 thing last year(https://github.com/layer5io/layer5/pull/1767), but this week might be a bit difficult as our team at Red Hat is on an office trip and I am in Pune as of now. I can take this up the following one if that would be fine

CC @leecalcote

**Comment by Nikhil-Ladha** [comment]
Sure, you can update us with that next week 👍🏼

**Comment by l5io** [comment]
🚀 Preview for commit 6112b01ab432f231bd588b8b6f0c27509b3fa078 at: https://62e84a592f372b3635635b9f--layer5.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 96f5f5108d31a1afd2dce835c5bc07f6a69512d6 at: https://62e8ec715ed8aa2fd73cac22--layer5.netlify.app

File: .github/workflows/lighthouse.yml (hunk 1 above)
Comment (Nikhil-Ladha) [reply .github/workflows/lighthouse.yml:28]: You can use the `find-comment` action, only if you have the PR number for which the workflow is running.
Or, if that's not possible then another way is to save the deploy URL as an artifact in the `preview-site` workflow and then download that artifact here and get the URL from it.

If the first one is feasible, then that is preferred.

**Comment by l5io** [comment]
🚀 Preview for commit 62418fb3e1158e75eb1b19a3b6fddad9d35cafc0 at: https://62ea00c04c8fb35442ba88b9--layer5.netlify.app

**Review by Chadha93 — COMMENTED** [review]
Lots's of great stuff here by @adithyaakrishna 💪. Good to go @Nikhil-Ladha @adithyaakrishna?

File: gatsby-config.js [hunk 4]
```diff
@@ -11,6 +11,7 @@ module.exports = {
     twitterUsername: "@layer5",
   },
   flags: {
+    PARALLEL_SOURCING: true,
     FAST_DEV: true,
```
Comment (Nikhil-Ladha) [inline gatsby-config.js:15]: It is enabled by default in gatsbyv4. Not needed

**Review by Nikhil-Ladha — COMMENTED** [review]
Also, the docker.png file that you have added is not needed

**Comment by Nikhil-Ladha** [comment]
> Lots's of great stuff here by @adithyaakrishna 💪. Good to go @Nikhil-Ladha @adithyaakrishna?

I doubt, the lighthouse.yml workflow is still incomplete

**Comment by adithyaakrishna** [comment]
@Nikhil-Ladha Yes, this is not yet complete, I am trying to work on a solution to find the preview links every time a code is pushed and this workflow is something we use at Red Hat as well. So I thought to implement something similar here

Ref: https://github.com/patternfly/patternfly-elements/blob/main/.github/workflows/visual-regression.yml#L65

**Comment by Nikhil-Ladha** [comment]
Can we expect an update here @adithyaakrishna ?

**Pull request closed without merging**